src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs

```diff
@@ -163,51 +161,56 @@ internal static IOpenApiSchema ResolveReferenceForSchema(OpenApiDocument documen
             {
-                schema.Properties[property.Key] = ResolveReferenceForSchema(document, property.Value);
+                schema.Properties[property.Key] = ResolveReferenceForSchema(document, property.Value, rootSchemaId);
             }
         }
 
         if (schema.AllOf is { Count: > 0 })
         {
             for (var i = 0; i < schema.AllOf.Count; i++)
             {
-                schema.AllOf[i] = ResolveReferenceForSchema(document, schema.AllOf[i]);
+                schema.AllOf[i] = ResolveReferenceForSchema(document, schema.AllOf[i], rootSchemaId);
             }
         }
 
         if (schema.OneOf is { Count: > 0 })
         {
             for (var i = 0; i < schema.OneOf.Count; i++)
             {
-                schema.OneOf[i] = ResolveReferenceForSchema(document, schema.OneOf[i]);
+                schema.OneOf[i] = ResolveReferenceForSchema(document, schema.OneOf[i], rootSchemaId);
             }
         }
 
         if (schema.AdditionalProperties is not null)
         {
-            schema.AdditionalProperties = ResolveReferenceForSchema(document, schema.AdditionalProperties);
+            schema.AdditionalProperties = ResolveReferenceForSchema(document, schema.AdditionalProperties, rootSchemaId);
         }
 
         if (schema.Items is not null)
         {
-            schema.Items = ResolveReferenceForSchema(document, schema.Items);
+            schema.Items = ResolveReferenceForSchema(document, schema.Items, rootSchemaId);
         }
 
         if (schema.Not is not null)
         {
-            schema.Not = ResolveReferenceForSchema(document, schema.Not);
+            schema.Not = ResolveReferenceForSchema(document, schema.Not, rootSchemaId);
         }
 
         // Handle schemas where the references have been inlined by the JsonSchemaExporter. In this case,
         // the `#` ID is generated by the exporter since it has no base document to baseline against. In this
         // case we we want to replace the reference ID with the schema ID that was generated by the
         // `CreateSchemaReferenceId` method in the OpenApiSchemaService.
         if (schema.Metadata is not null &&
-            schema.Metadata.ContainsKey(OpenApiConstants.RefId) &&
-            schema.Metadata.TryGetValue(OpenApiConstants.SchemaId, out var schemaId) &&
-            schemaId is string schemaIdString)
+            schema.Metadata.TryGetValue(OpenApiConstants.RefId, out var refId) &&
+            refId is string refIdString)
         {
-            return document.AddOpenApiSchemaByReference(schemaIdString, schema);
+            if (schema.Metadata.TryGetValue(OpenApiConstants.SchemaId, out var schemaId) &&
+                schemaId is string schemaIdString)
+            {
+                return document.AddOpenApiSchemaByReference(schemaIdString, schema);
+            }
+            var relativeSchemaId = $"#/components/schemas/{rootSchemaId}{refIdString.Replace("#", string.Empty)}";
```

Comment:
Is this where the `tags` and `items` updates in the tests comes from?

Do we want to add any explicit testing here?

Reply:
Yep, this addresses at least one of the cases now that OpenApi.NET supports resolving relative JSON schemas references. There's still one remaining test case around this issue that s commented it out that we are working on.